<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using stackr to run stacks pipeline inside R/RStudio • stackr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using stackr to run stacks pipeline inside R/RStudio">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stackr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/stackr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/thierrygosselin/stackr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Using stackr to run stacks pipeline inside R/RStudio</h1>
                        <h4 class="author">Thierry Gosselin</h4>
            
            <h4 class="date">2019-12-03</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/stackr/blob/master/vignettes/stackr.Rmd"><code>vignettes/stackr.Rmd</code></a></small>
      <div class="hidden name"><code>stackr.Rmd</code></div>

    </div>

    
    
<p>This vignette purposes is to show users how to use stackr to run the <a href="http://catchenlab.life.illinois.edu/stacks/">stacks</a> pipeline inside R/RStudio.</p>
<div id="stackr-assumptions" class="section level1">
<h1 class="hasAnchor">
<a href="#stackr-assumptions" class="anchor"></a>stackr assumptions</h1>
<ul>
<li>stackr won’t make your stacks data better. Bad input data is just that… bad!</li>
<li>You have read <a href="http://catchenlab.life.illinois.edu/stacks/manual/">stacks manual</a> and understand the separate module and the output.</li>
<li>You have read the associated stacks paper. The latest one is:</li>
</ul>
<p>Rochette, N., Rivera‐Colón, A., Catchen, J. (2019). Stacks 2: Analytical methods for paired‐end sequencing improve RADseq‐based population genomics. Molecular Ecology <a href="https://dx.doi.org/10.1111/mec.15253" class="uri">https://dx.doi.org/10.1111/mec.15253</a></p>
</div>
<div id="stackr-advantages" class="section level1">
<h1 class="hasAnchor">
<a href="#stackr-advantages" class="anchor"></a>stackr advantages</h1>
<ul>
<li>Allows to run stacks inside R/RStudio/RStudio Server.</li>
<li>The philosophy of working by project with pre-organized folders.</li>
<li>Some important steps are <strong>parallelized</strong>.</li>
<li>You have more than 1 sequencing chip/lane ? This workflow will save you lots of time.</li>
<li>
<strong>Technical replicates</strong>, inside or across chip/lanes are managed uniquely.</li>
<li>
<strong>nightmares because of a crashed computer/cluster/server?</strong> stackr manage stacks unique integer (previously called SQL IDs) throughout the pipeline. It’s integrated from the start, making it a breeze to just re-start your pipeline after a crash!</li>
<li>
<strong>mismatch testing:</strong> <em>de novo</em> mismatch threshold series is integrated inside <code>run_ustacks</code> and stackr will produce tables and figures automatically.</li>
<li>
<strong>catalog</strong>: for bigger sampling size project, breaking down the catalog into several separate <em>cstacks</em> steps makes the pipeline more rigorous if your computer/cluster/server crash.</li>
<li>
<strong>logs</strong> generated by stacks are read and transferred in human-readable tables/tibbles. Detecting problems is easier.</li>
<li>summary of different stacks modules: available automatically inside stackr pipeline, but also available for users who didn’t use stackr to run stacks.</li>
<li>Detecting problems is easier.</li>
<li>Increased reproducibly.</li>
</ul>
</div>
<div id="coding-style" class="section level1">
<h1 class="hasAnchor">
<a href="#coding-style" class="anchor"></a>Coding style</h1>
<ul>
<li>stackr is trying hard to follow the <a href="https://style.tidyverse.org">tidyverse style guide</a>.</li>
<li>inside stacks, parallel execution and the use of threads sometimes uses the argument <code>-p</code>, <code>-t</code>, <code>--threads</code>. Depending on module the meaning changes, which can be confusing or generate problems for first-time users. Inside stackr, using parallel execution is always through the use of the argument: <code>parallel.core</code>.</li>
<li>argument break down inside stacks modules sometimes uses the hyphen <code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code>. Inside stackr this is replaced by function arguments with <code>.</code> point/dot. Example: in <strong>stacks</strong> <code>process_radtags</code> argument <code>--adapter-1</code> will be the argument <code>adapter.1</code> inside <strong>stackr</strong> <code>run_process_radtags</code>.</li>
</ul>
</div>
<div id="workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#workflow" class="anchor"></a>Workflow</h1>
<p>The names of stackr functions follows stacks module, below the similarity:</p>
<p><img src="../../../../../Mobile%20Documents/com~apple~CloudDocs/r_packages/stackr/vignettes/stackr_workflow.png" width="570" style="display: block; margin: auto;"></p>
<div id="dependencies" class="section level2">
<h2 class="hasAnchor">
<a href="#dependencies" class="anchor"></a>Dependencies</h2>
<ul>
<li>Make sure you have the latest R and RStudio version before installing stackr. My other package <a href="https://thierrygosselin.github.io/radiator/index.html">radiator</a> as a nice <a href="https://thierrygosselin.github.io/radiator/articles/rad_genomics_computer_setup.html">vignette</a> for this.</li>
<li>Install <a href="http://catchenlab.life.illinois.edu/stacks/manual/#install">stacks</a> and it’s dependencies (my own installation pipeline <a href="https://thierrygosselin.github.io/RADseq_cloud_tutorial/radseq_cloud_tutorial.html#stacks">here</a>).</li>
<li><strong>don’t write to me if you’re having problem at this step.</strong></li>
<li>
<a href="https://groups.google.com/forum/?fromgroups#!forum/stacks-users">stacks google group</a> is your best friend.</li>
<li>Install stackr by following the <a href="https://github.com/thierrygosselin/stackr#installation">instructions</a>.</li>
</ul>
</div>
<div id="generate-project-folders" class="section level2">
<h2 class="hasAnchor">
<a href="#generate-project-folders" class="anchor"></a>Generate project folders</h2>
<p>Here, I’m going to use RStudio, because I really got to like it and I’m very efficient in it. But it’s by no mean essential.</p>
<ol style="list-style-type: decimal">
<li><p>Set your working directory inside RStudio using <code>setwd</code> function and appropriate path.</p></li>
<li><p>Generate the folders for your project</p></li>
</ol>
<p>To get the doc on the function we’re about to use: <code><a href="../reference/build_stackr_workflow_dir.html">?stackr::build_stackr_workflow_dir</a></code>.</p>
<p>Here it’s a project on turtle…</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">stackr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/stackr/man/build_stackr_workflow_dir.html">build_stackr_workflow_dir</a></span>(<span class="dt">main.folder.name =</span> <span class="st">"turtle"</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="co"># stackr::build_stackr_workflow_dir(main.folder.name = "turtle", date = FALSE) #if you don't want the date appended</span></a></code></pre></div>
<p>By default, the date and time is appended to the main folder and inside, these folders will be generated:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="dv">01</span>_scripts</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="dv">02</span>_project_info</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="dv">03</span>_sequencing_lanes</a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="dv">04</span>_process_radtags</a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="dv">05</span>_clustering_mismatches</a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="dv">06</span>_ustacks_<span class="dv">2</span>_gstacks</a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="dv">07</span>_populations</a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="dv">08</span>_stacks_results</a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="dv">09</span>_log_files</a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="dv">10</span>_filters</a></code></pre></div>
<p>For now let’s just say that you put the scripts and other <code>.R</code> files associated to this project inside the first folder: <code>01_scripts</code>.</p>
</div>
<div id="start-a-project" class="section level2">
<h2 class="hasAnchor">
<a href="#start-a-project" class="anchor"></a>Start a project</h2>
<p>Inside RStudio generate a new project inside the directory just created:</p>
<p><img src="../../../../../Mobile%20Documents/com~apple~CloudDocs/r_packages/stackr/vignettes/rstudio.project.1.png" width="564" style="display: block; margin: auto;"></p>
<p><img src="../../../../../Mobile%20Documents/com~apple~CloudDocs/r_packages/stackr/vignettes/rstudio.project.2.png" width="570" style="display: block; margin: auto;"></p>
<p>You can add the alias/shorcut for the project on your computer desktop.</p>
</div>
<div id="project-info-file" class="section level2">
<h2 class="hasAnchor">
<a href="#project-info-file" class="anchor"></a>Project info file</h2>
<p>Inside MS Excel or another software that works with tabular columns, prepare your project info file and save it with the file extension <code>.tsv</code> inside the second folder: <code>02_project_info</code>. For example, the file <code>project.info.turtle.tsv</code> will contain a minimum of 3 columns for <strong>single-end</strong> and 4 columns for <strong>paired-end</strong> sequencing. The number of lines equal the number of sample: 1 sample per line. The project info file is <strong>sample</strong> oriented or focused.</p>
<p>But before generating and filling the project info file consider the names of your samples…</p>
<div id="the-art-of-naming-your-samples" class="section level3">
<h3 class="hasAnchor">
<a href="#the-art-of-naming-your-samples" class="anchor"></a>The art of naming your samples</h3>
<p>You already have somewhere on your computer barcodes and sample name. Here is the time to reconsider how you named your sample. This is not trivial, researcher loose a lot of time because of bad naming scheme. REALLY.</p>
<p><strong>My philosophy:</strong></p>
<p>Include useful metadata associated with your sample separated by an <a href="https://en.wikipedia.org/wiki/Hyphen">hyphen</a> <code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code>. Why an hyphen <code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code> instead of a point <code>.</code> or an underscore <code>_</code> ?? Imagine you print your file on real paper or that you have to translate by hand the sample names. An hyphen, when printed, generated far less confusion because it can easily be distinguised if printed on paper (lined paper or not). The underscore or point will just be confusion with space.</p>
<ul>
<li>Please don’t use space</li>
<li>Please use ALL CAPS so that confusion is minimized.</li>
<li>The key here is consistency.</li>
</ul>
<p>Below an example using metadata: <code>SPECIES</code>, <code>STRATA</code> (could be location, I don’t know the pop yet), <code>YEAR</code> (of sampling), <code>SEX</code>, <code>ID</code> (with sufficient digits)</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">STU<span class="op">-</span>BUR<span class="dv">-2019</span><span class="op">-</span>F<span class="dv">-0001</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">STU<span class="op">-</span>BUR<span class="dv">-2019</span><span class="op">-</span>F<span class="dv">-0002</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3">STU<span class="op">-</span>BUR<span class="dv">-2019</span><span class="op">-</span>F<span class="dv">-0003</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4">STU<span class="op">-</span>BUR<span class="dv">-2019</span><span class="op">-</span>F<span class="dv">-0004</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5">STU<span class="op">-</span>BUR<span class="dv">-2019</span><span class="op">-</span>M<span class="dv">-0005</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6">STU<span class="op">-</span>BUR<span class="dv">-2017</span><span class="op">-</span>F<span class="dv">-0006</span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7">STU<span class="op">-</span>GRA<span class="dv">-2017</span><span class="op">-</span>M<span class="dv">-0007</span> <span class="co"># you could start over from 0001 with different strata, it doesn't matter</span></a>
<a class="sourceLine" id="cb3-8" data-line-number="8">STU<span class="op">-</span>GRA<span class="dv">-2018</span><span class="op">-</span>M<span class="dv">-0008</span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9">STU<span class="op">-</span>GRA<span class="dv">-2019</span><span class="op">-</span>M<span class="dv">-0009</span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10">STU<span class="op">-</span>GRA<span class="dv">-2019</span><span class="op">-</span>F<span class="dv">-0010</span></a>
<a class="sourceLine" id="cb3-11" data-line-number="11">STU<span class="op">-</span>GRA<span class="dv">-2019</span><span class="op">-</span>F<span class="dv">-0011</span></a></code></pre></div>
<p><strong>Special considerations:</strong></p>
<ul>
<li>keep it simple.</li>
<li>you can use 3 or 4 letters for the metadata, but inside 1 metadata but please please <strong>be consistent</strong>.</li>
<li>future parsing, if you need it, will be very very easy this way.</li>
</ul>
<p>Ok now the project info file…</p>
</div>
<div id="single-end" class="section level3">
<h3 class="hasAnchor">
<a href="#single-end" class="anchor"></a>single-end</h3>
<p><strong>Columns prerequisites:</strong></p>
<ul>
<li>
<code>LANES</code>: the name of the sequencing file e.g. H6CY7BBXY_3_254.fastq.gz</li>
<li>
<code>BARCODES</code>: the barcode associated with the sample.</li>
<li>
<code>INDIVIDUALS</code>: a more appealing sample name (more on this below).</li>
</ul>
<p>Generate an empty file like this:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">tibble<span class="op">::</span><span class="kw"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span>(</a>
<a class="sourceLine" id="cb4-2" data-line-number="2">  <span class="dt">LANES =</span> <span class="kw"><a href="https://rdrr.io/r/base/character.html">character</a></span>(<span class="dv">0</span>), </a>
<a class="sourceLine" id="cb4-3" data-line-number="3">  <span class="dt">BARCODES =</span> <span class="kw"><a href="https://rdrr.io/r/base/character.html">character</a></span>(<span class="dv">0</span>), </a>
<a class="sourceLine" id="cb4-4" data-line-number="4">  <span class="dt">INDIVIDUALS =</span> <span class="kw"><a href="https://rdrr.io/r/base/character.html">character</a></span>(<span class="dv">0</span>)</a>
<a class="sourceLine" id="cb4-5" data-line-number="5">) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="st">  </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_tsv</a></span>(<span class="dt">x =</span> ., <span class="dt">path =</span> <span class="st">"02_project_info/project.info.turtle.tsv"</span>)</a></code></pre></div>
</div>
<div id="paired-end" class="section level3">
<h3 class="hasAnchor">
<a href="#paired-end" class="anchor"></a>paired-end</h3>
<p><strong>Columns prerequisites:</strong></p>
<ul>
<li>
<code>BARCODES</code>: the barcode associated with the sample</li>
<li>
<code>INDIVIDUALS</code>: a more appealing sample name (more on this below)</li>
<li>
<code>FORWARD</code>: the name of the forward sequencing file e.g. H5CY5BBXY_2_25_1.fastq.gz</li>
<li>
<code>REVERSE</code>: the name of the reverse sequencing file e.g. H5CY5BBXY_2_25_2.fastq.gz</li>
</ul>
<p>Generate an empty file like this:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">tibble<span class="op">::</span><span class="kw"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span>(</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">  <span class="dt">BARCODES =</span> <span class="kw"><a href="https://rdrr.io/r/base/character.html">character</a></span>(<span class="dv">0</span>), </a>
<a class="sourceLine" id="cb5-3" data-line-number="3">  <span class="dt">INDIVIDUALS =</span> <span class="kw"><a href="https://rdrr.io/r/base/character.html">character</a></span>(<span class="dv">0</span>),</a>
<a class="sourceLine" id="cb5-4" data-line-number="4">  <span class="dt">FORWARD =</span> <span class="kw"><a href="https://rdrr.io/r/base/character.html">character</a></span>(<span class="dv">0</span>),</a>
<a class="sourceLine" id="cb5-5" data-line-number="5">  <span class="dt">REVERSE =</span> <span class="kw"><a href="https://rdrr.io/r/base/character.html">character</a></span>(<span class="dv">0</span>)</a>
<a class="sourceLine" id="cb5-6" data-line-number="6">) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="st">  </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_tsv</a></span>(<span class="dt">x =</span> ., <span class="dt">path =</span> <span class="st">"02_project_info/project.info.turtle.tsv"</span>)</a></code></pre></div>
</div>
<div id="faq" class="section level3">
<h3 class="hasAnchor">
<a href="#faq" class="anchor"></a>FAQ</h3>
<div id="multiple-sequencing-lanes-or-chips" class="section level4">
<h4 class="hasAnchor">
<a href="#multiple-sequencing-lanes-or-chips" class="anchor"></a>Multiple sequencing lanes or chips ?</h4>
<p>If you have several Illumina lanes or Ion Torrent chips, just write them down along the sample and barcodes associated with it. On the same file.</p>
</div>
<div id="technical-replicates" class="section level4">
<h4 class="hasAnchor">
<a href="#technical-replicates" class="anchor"></a>Technical replicates ?</h4>
<p>2 options:</p>
<ul>
<li>use separate name for it or best,</li>
<li>let stackr manage it my using the same name (yes… yes…).</li>
</ul>
<p>stackr will see that you have, for different barcodes/lanes combo, ids that have the exact same name. What will happen ?</p>
<ul>
<li>things will be kept separate and <code>-1</code> and <code>-2</code> will be added to the sample name.</li>
<li>after <code>run_process_radtags</code>, the replicates will be combined and the new file will have <code>-R</code> at the end. The individual replicates are kept separate as well.</li>
</ul>
<p><strong>example:</strong></p>
<p>You have a sample, <code>STU-BUR-2019-F-0001</code>, on 2 separate wells on the the plate or even lanes, with different barcode. At the end of <code>run_process_radtags</code> you’ll have:</p>
<ul>
<li><code>STU-BUR-2019-F-0001-1</code></li>
<li><code>STU-BUR-2019-F-0001-2</code></li>
<li>
<code>STU-BUR-2019-F-0001-R</code> (for the combined replicates)</li>
</ul>
</div>
<div id="no-lanes-just-individuals-fq-files" class="section level4">
<h4 class="hasAnchor">
<a href="#no-lanes-just-individuals-fq-files" class="anchor"></a>No lanes just individual’s fq files ?</h4>
<p>Example: you used DArT sequencing services. You have individual fq files, the barcodes are still in there along some Illumina adapters.</p>
<p>Use the <code>LANES</code> column for the individual FQ files that requires to go through stacks process_radtags.</p>
</div>
</div>
</div>
<div id="transfer-sequencing-files" class="section level2">
<h2 class="hasAnchor">
<a href="#transfer-sequencing-files" class="anchor"></a>Transfer sequencing files</h2>
<p>Transfer the sequencing files inside the folder: <code>03_sequencing_lanes</code>.</p>
</div>
<div id="quality-control" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-control" class="anchor"></a>Quality control</h2>
<p>I highly recommend running the sequencing files through <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>. If you have bad sequencing, no need to go further in the pipeline…</p>
</div>
<div id="run_process_radtags" class="section level2">
<h2 class="hasAnchor">
<a href="#run_process_radtags" class="anchor"></a>run_process_radtags</h2>
<p>This is the demultiplexing step in stacks, the module is called <a href="http://catchenlab.life.illinois.edu/stacks/comp/process_radtags.php">process_radtags</a>.</p>
<p>Function documentation: <code><a href="../reference/run_process_radtags.html">?stackr::run_process_radtags</a></code> (read it twice).</p>
<p>Difference with stacks <code>process_radtags</code> ?</p>
<ul>
<li>You have one function that will run all you lanes/chips.</li>
<li>Technical replicates will be managed and combined.</li>
<li>A summary of the process will be generated.</li>
</ul>
<p>The function will generate, inside <code>08_stacks_results</code>, files like these:</p>
<ul>
<li><code>process_radtags_LANE_01_barcodes_stats.tsv</code></li>
<li><code>process_radtags_LANE_01_stats.tsv</code></li>
</ul>
<p>Containing statistics per barcodes and per lanes. The logs are found inside the folder: <code>09_log_files</code>.</p>
<p>The project info file will be populated with these columns/info (example below using paire-end data): * <code>TOTAL</code>: the number of total reads. * <code>NO_RADTAG</code>: the number of reads without radtags. * <code>LOW_QUALITY</code>: the number of reads discarded because of low quality. * <code>RETAINED</code>: The number of reads retained after the filters used inside process_radtags. * <code>SQL_ID</code>: A unique integer that will be used inside the next step: <code>run_ustacks</code> * <code>FQ_FILES_F</code>: the individual forward fq files * <code>FQ_FILES_R</code>: the individual reverse fq files</p>
<div id="example-single-end-2-enzymes-radseqgbs-data" class="section level3">
<h3 class="hasAnchor">
<a href="#example-single-end-2-enzymes-radseqgbs-data" class="anchor"></a>Example: single-end, 2 enzymes RADseq/GBS data</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">step1 &lt;-<span class="st"> </span>stackr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/stackr/man/run_process_radtags.html">run_process_radtags</a></span>(</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">  <span class="dt">project.info =</span> <span class="st">"02_project_info/project.info.turtle.tsv"</span>,</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">  <span class="dt">path.seq.lanes =</span> <span class="st">"03_sequencing_lanes"</span>,</a>
<a class="sourceLine" id="cb6-4" data-line-number="4">  <span class="dt">q =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb6-5" data-line-number="5">  <span class="dt">renz.1 =</span> <span class="st">"pstI"</span>, </a>
<a class="sourceLine" id="cb6-6" data-line-number="6">  <span class="dt">renz.2 =</span> <span class="st">"sphI"</span>,</a>
<a class="sourceLine" id="cb6-7" data-line-number="7">  <span class="dt">t =</span> <span class="dv">90</span>, </a>
<a class="sourceLine" id="cb6-8" data-line-number="8">  <span class="dt">len.limit =</span> <span class="dv">90</span>,</a>
<a class="sourceLine" id="cb6-9" data-line-number="9">  <span class="dt">filter.illumina =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb6-10" data-line-number="10">  <span class="dt">retain.header =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb6-11" data-line-number="11">  <span class="dt">adapter.1 =</span> <span class="st">"CGAGATCGGAAGAGCGGG"</span>, </a>
<a class="sourceLine" id="cb6-12" data-line-number="12">  <span class="dt">adapter.mm =</span> <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb6-13" data-line-number="13">  <span class="dt">parallel.core =</span> <span class="dv">12</span></a>
<a class="sourceLine" id="cb6-14" data-line-number="14">)</a>
<a class="sourceLine" id="cb6-15" data-line-number="15"><span class="co"># look for the remaining default arguments...</span></a></code></pre></div>
</div>
<div id="example-paired-end-1-enzyme-radseq" class="section level3">
<h3 class="hasAnchor">
<a href="#example-paired-end-1-enzyme-radseq" class="anchor"></a>Example: paired-end 1 enzyme RADseq</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">step1 &lt;-<span class="st"> </span>stackr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/stackr/man/run_process_radtags.html">run_process_radtags</a></span>(</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">  <span class="dt">P =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">  <span class="dt">project.info =</span> <span class="st">'02_project_info/project.info.turtle.tsv'</span>,</a>
<a class="sourceLine" id="cb7-4" data-line-number="4">  <span class="dt">path.seq.lanes =</span> <span class="st">"03_sequencing_lanes"</span>,</a>
<a class="sourceLine" id="cb7-5" data-line-number="5">  <span class="dt">enzyme =</span> <span class="st">"sbfI"</span>,</a>
<a class="sourceLine" id="cb7-6" data-line-number="6">  <span class="dt">adapter.1 =</span> <span class="st">"CGAGATCGGAAGAGCGGG"</span>,</a>
<a class="sourceLine" id="cb7-7" data-line-number="7">  <span class="dt">adapter.2 =</span> <span class="st">"CGAGATCGGAAGAGCGGG"</span>,</a>
<a class="sourceLine" id="cb7-8" data-line-number="8">  <span class="dt">adapter.mm =</span> <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb7-9" data-line-number="9">  <span class="dt">t =</span> <span class="dv">90</span>, </a>
<a class="sourceLine" id="cb7-10" data-line-number="10">  <span class="dt">len.limit =</span> <span class="dv">90</span>,</a>
<a class="sourceLine" id="cb7-11" data-line-number="11">  <span class="dt">retain.header =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb7-12" data-line-number="12">  <span class="dt">parallel.core =</span> <span class="dv">8</span></a>
<a class="sourceLine" id="cb7-13" data-line-number="13">)</a>
<a class="sourceLine" id="cb7-14" data-line-number="14"><span class="co"># look for the remaining default arguments...</span></a></code></pre></div>
</div>
</div>
<div id="run_ustacks" class="section level2">
<h2 class="hasAnchor">
<a href="#run_ustacks" class="anchor"></a>run_ustacks</h2>
<p>This is the <em>de novo</em> assembly step in stacks, the module is called <a href="http://catchenlab.life.illinois.edu/stacks/comp/ustacks.php">ustacks</a>.</p>
<p>Function documentation: <code><a href="../reference/run_ustacks.html">?stackr::run_ustacks</a></code> (read it twice).</p>
<p>Difference with stacks <code>ustacks</code> ?</p>
<ul>
<li>SQL IDs or the <code>-i</code> argument, the unique integer ID for this sample is managed automatically.</li>
<li>computer/cluster/server crash ? Just re-run the function…</li>
<li>The mismatch testing can be done inside <code><a href="../reference/run_ustacks.html">stackr::run_ustacks</a></code> automatically ;)</li>
<li>A summary is generated and can be run separately for those who ran the function inside the terminal with stacks. Documentation: <code><a href="../reference/summary_ustacks.html">?summary_ustacks</a></code>.</li>
</ul>
<p>If you don’t know the optimal number of mismatches between your individual reads this part is the most important steps of <em>de novo</em> assembly.</p>
<div id="mismatches" class="section level3">
<h3 class="hasAnchor">
<a href="#mismatches" class="anchor"></a>Mismatches</h3>
<p>When using paired-end data, use forward reads for the analysis.</p>
<p>Select one sample to do the test. For this, you could use a sample with average file size or use the stats produced directly by the output of <code>run_process_radtags</code> and available inside the updated project info file saved by the function.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">info.stats &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span>(<span class="dt">file =</span> <span class="st">"02_project_info/project.info.20191121@1606.tsv"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(info.stats, REPLICATES <span class="op">==</span><span class="st"> "1"</span>)</a>
<a class="sourceLine" id="cb8-3" data-line-number="3">m &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(info.stats<span class="op">$</span>RETAINED)</a>
<a class="sourceLine" id="cb8-4" data-line-number="4">s &lt;-<span class="st"> </span>stats<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span>(info.stats<span class="op">$</span>RETAINED)</a>
<a class="sourceLine" id="cb8-5" data-line-number="5"></a>
<a class="sourceLine" id="cb8-6" data-line-number="6">read.min &lt;-<span class="st"> </span>m <span class="op">-</span><span class="st"> </span>s</a>
<a class="sourceLine" id="cb8-7" data-line-number="7">read.max &lt;-<span class="st"> </span>m <span class="op">+</span><span class="st"> </span>s</a>
<a class="sourceLine" id="cb8-8" data-line-number="8"></a>
<a class="sourceLine" id="cb8-9" data-line-number="9">sample.select &lt;-<span class="st"> </span>info.stats <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-10" data-line-number="10"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(RETAINED <span class="op">&gt;=</span><span class="st"> </span>read.min <span class="op">&amp;</span><span class="st"> </span>RETAINED <span class="op">&lt;=</span><span class="st"> </span>read.max) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-11" data-line-number="11"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/sample.html">sample_n</a></span>(<span class="dt">size =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb8-12" data-line-number="12">sample.select<span class="op">$</span>FQ_FILES_F</a>
<a class="sourceLine" id="cb8-13" data-line-number="13"><span class="co"># [1] "BET-LAM-60044-1682200-1.1.fq.gz"</span></a></code></pre></div>
<p>Make a copy the file inside the folder: <code>05_clustering_mismatches</code>:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/files.html">file.copy</a></span>(</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">  <span class="dt">from =</span> <span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="st">"04_process_radtags"</span>, sample.select<span class="op">$</span>FQ_FILES_F), </a>
<a class="sourceLine" id="cb9-3" data-line-number="3">  <span class="dt">to =</span> <span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="st">"05_clustering_mismatches/"</span>, sample.select<span class="op">$</span>FQ_FILES_F)</a>
<a class="sourceLine" id="cb9-4" data-line-number="4">)</a></code></pre></div>
<p><strong>Run the mismatch clustering series:</strong></p>
<ul>
<li>with files around 2M reads</li>
<li>testing 1 to 5 mismatches is very fast to run</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">mismatch &lt;-<span class="st"> </span>stackr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/stackr/man/run_ustacks.html">run_ustacks</a></span>(</a>
<a class="sourceLine" id="cb10-2" data-line-number="2">  <span class="dt">mismatch.testing =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb10-3" data-line-number="3">  <span class="dt">f =</span> <span class="st">"05_clustering_mismatches"</span>,</a>
<a class="sourceLine" id="cb10-4" data-line-number="4">  <span class="dt">M =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,</a>
<a class="sourceLine" id="cb10-5" data-line-number="5">  <span class="dt">model.type =</span> <span class="st">"bounded"</span>)</a></code></pre></div>
<p>This will generate 5 folders inside <code>05_clustering_mismatches</code> that will have the mismatches number and associated ustacks output files in it.</p>
<p>A figure similar to this one:</p>
<p><img src="../../../../../Mobile%20Documents/com~apple~CloudDocs/r_packages/stackr/vignettes/mismatches.plot.png" width="1546" style="display: block; margin: auto;"></p>
<p>To understand this figure, I suggest you read these 2 excellent papers:</p>
<p>Ilut, D., Nydam, M., Hare, M. (2014). Defining Loci in Restriction-Based Reduced Representation Genomic Data from Non model Species: Sources of Bias and Diagnostics for Optimal Clustering BioMed Research International 2014(2), 1 9. <a href="https://dx.doi.org/10.1155/2014/675158" class="uri">https://dx.doi.org/10.1155/2014/675158</a></p>
<p>Harvey, M., Judy, C., Seeholzer, G., Maley, J. (2015). Similarity thresholds used in short read assembly reduce the comparability of population histories across species PeerJ 3(), e895. <a href="https://dx.doi.org/10.7717/peerj.895" class="uri">https://dx.doi.org/10.7717/peerj.895</a></p>
<p>The function also produce a tibble <code>mismatches.summary.tsv</code> with several columns and lines summarizing ustacks output logs for the different mismatches.</p>
<p><img src="../../../../../Mobile%20Documents/com~apple~CloudDocs/r_packages/stackr/vignettes/mismatches.summary.1.png" width="680" style="display: block; margin: auto;"></p>
<p>This is the data used to make the figure above.</p>
<p><img src="../../../../../Mobile%20Documents/com~apple~CloudDocs/r_packages/stackr/vignettes/mismatches.summary.2.png" width="659" style="display: block; margin: auto;"></p>
</div>
<div id="de-novo-assembly" class="section level3">
<h3 class="hasAnchor">
<a href="#de-novo-assembly" class="anchor"></a>de novo assembly</h3>
<p>Now that you have decided the mismatch threshold, corresponding to the <code>M</code> argument inside <code><a href="../reference/run_ustacks.html">stackr::run_ustacks</a></code>, run the function:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">u &lt;-<span class="st"> </span>stackr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/stackr/man/run_ustacks.html">run_ustacks</a></span>(</a>
<a class="sourceLine" id="cb11-2" data-line-number="2">  <span class="dt">project.info =</span> <span class="st">"02_project_info/project.info.20191121@1606.tsv"</span>,</a>
<a class="sourceLine" id="cb11-3" data-line-number="3">  <span class="dt">M =</span> <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb11-4" data-line-number="4">  <span class="dt">model.type =</span> <span class="st">"bounded"</span>,</a>
<a class="sourceLine" id="cb11-5" data-line-number="5">  <span class="dt">parallel.core =</span> <span class="dv">12</span></a>
<a class="sourceLine" id="cb11-6" data-line-number="6">)</a>
<a class="sourceLine" id="cb11-7" data-line-number="7"><span class="co"># careful with the defaults...</span></a></code></pre></div>
<p>When using paired-end data, the function uses the forward reads for the assembly. The reverse reads are used later in the pipeline.</p>
<p><strong>Output: </strong></p>
<p>The folder <code>06_ustacks_2_gstacks</code> will be populated with 3 files per samples ending with:</p>
<ul>
<li><code>alleles.tsv.gz</code></li>
<li><code>tags.tsv.gz</code></li>
<li><code>snps.tsv.gz</code></li>
</ul>
<p>These files are <a href="http://catchenlab.life.illinois.edu/stacks/manual/#files">documented here</a></p>
<p>The folder <code>08_stacks_results</code> will have a summary file: <code>summary_ustacks_date@time.tsv</code>.</p>
<p>The folder <code>09_log_files</code> as the associated individual log files.</p>
</div>
</div>
<div id="run_cstacks" class="section level2">
<h2 class="hasAnchor">
<a href="#run_cstacks" class="anchor"></a>run_cstacks</h2>
<p>This is the catalog or pseudo reference genome for the <em>de novo</em> part of stacks, the module is called <a href="http://catchenlab.life.illinois.edu/stacks/comp/cstacks.php">cstacks</a>.</p>
<p>Function documentation: <code><a href="../reference/run_cstacks.html">?stackr::run_cstacks</a></code> (read it twice).</p>
<p>Difference with stacks <code>cstacks</code> ?</p>
<ul>
<li>Generating the catalog is split in x number of samples to make it more robust.</li>
</ul>
<p><strong>How?</strong></p>
<ul>
<li>The population map that shows which samples to use to build the catalog is split in 20 individuals (by default).</li>
<li>The function generates a catalog after the first 20 samples and writes it down.</li>
<li>
<code>cstacks</code> starts again with the previous catalog and build a new one with new samples next on the list.</li>
<li>It’s incremental</li>
<li>Saves you a lot of time if e.g. your computer crash.</li>
<li>You can easily start with the catalog that was last written in the folder.</li>
</ul>
<p>For now, I highly discourage subsampling to build the catalog = please use all the samples in the project.</p>
<p>Start by generating a population map:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">pop.map.catalog &lt;-<span class="st"> </span>u<span class="op">$</span>project.info <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/distinct.html">distinct</a></span>(INDIVIDUALS_REP) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-3" data-line-number="3"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">POP_ID =</span> 1L) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-4" data-line-number="4"><span class="st">  </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_tsv</a></span>(</a>
<a class="sourceLine" id="cb12-5" data-line-number="5">    <span class="dt">x =</span> .,</a>
<a class="sourceLine" id="cb12-6" data-line-number="6">    <span class="dt">path =</span> <span class="st">"02_project_info/population.map.turtle.catalog.tsv"</span>,</a>
<a class="sourceLine" id="cb12-7" data-line-number="7">    <span class="dt">col_names =</span> FALSE<span class="co"># stacks doesn't want column header</span></a>
<a class="sourceLine" id="cb12-8" data-line-number="8">  ) </a></code></pre></div>
<p><strong>Build the catalog:</strong></p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">c &lt;-<span class="st"> </span>stackr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/stackr/man/run_cstacks.html">run_cstacks</a></span> (</a>
<a class="sourceLine" id="cb13-2" data-line-number="2">  <span class="dt">P =</span> <span class="st">"06_ustacks_2_gstacks"</span>,</a>
<a class="sourceLine" id="cb13-3" data-line-number="3">  <span class="dt">M =</span> <span class="st">"02_project_info/population.map.turtle.catalog.tsv"</span>,</a>
<a class="sourceLine" id="cb13-4" data-line-number="4">  <span class="dt">n =</span> <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb13-5" data-line-number="5">  <span class="dt">parallel.core =</span> <span class="dv">32</span></a>
<a class="sourceLine" id="cb13-6" data-line-number="6">)</a></code></pre></div>
</div>
<div id="run_sstacks" class="section level2">
<h2 class="hasAnchor">
<a href="#run_sstacks" class="anchor"></a>run_sstacks</h2>
<p>That’s the <a href="http://catchenlab.life.illinois.edu/stacks/comp/sstacks.php">sstacks</a> part in stacks. It’s where the output from ustacks (the individual files) are matched against the catalog.</p>
<p>Function documentation: <code><a href="../reference/run_sstacks.html">?stackr::run_sstacks</a></code> (read it twice).</p>
<p>Difference with stacks <code>sstacks</code> ?</p>
<ul>
<li>Not much.</li>
<li>But I like the summary it does automatically at the end. It allows to see potential problem in the data.</li>
<li>Computer or server problem during the sstacks ? Just launch again the same command, the function will start again, but only with the unmatched samples!</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">s &lt;-<span class="st"> </span>stackr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/stackr/man/run_sstacks.html">run_sstacks</a></span>(<span class="dt">parallel.core =</span> <span class="dv">12</span>)</a></code></pre></div>
</div>
<div id="run_tsv2bam" class="section level2">
<h2 class="hasAnchor">
<a href="#run_tsv2bam" class="anchor"></a>run_tsv2bam</h2>
<p>That’s the <a href="http://catchenlab.life.illinois.edu/stacks/comp/tsv2bam.php">tsv2bam</a> part in stacks. The tsv2bam program will transpose data so that it is oriented by locus, instead of by sample. This allows downstream improvements (e.g. Bayesian SNP calling).</p>
<p>If paired-end reads are available, the <code>tsv2bam</code> program will pull in the set of paired-end reads that are associated with each single-end locus that was assembled <em>de novo</em>.</p>
<p>Difference with stacks <code>tsv2bam</code> ?</p>
<ul>
<li>after running this function, there’s no need to run <code>SAMtools</code>. It’s all integrated inside <code>run_tsv2bam</code>.</li>
<li>The function uses <code>SAMtools</code> to merge <strong>in parallel</strong> the BAM sample files into a unique BAM catalog file..</li>
</ul>
<p>Function documentation: <code><a href="../reference/run_tsv2bam.html">?stackr::run_tsv2bam</a></code> (read it twice).</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="co"># example with paired-end data:</span></a>
<a class="sourceLine" id="cb15-2" data-line-number="2">bam.sum &lt;-<span class="st"> </span>stackr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/stackr/man/run_tsv2bam.html">run_tsv2bam</a></span>(</a>
<a class="sourceLine" id="cb15-3" data-line-number="3">  <span class="dt">M =</span> <span class="st">"02_project_info/population.map.turtle.catalog.tsv"</span>,</a>
<a class="sourceLine" id="cb15-4" data-line-number="4">  <span class="dt">R =</span> <span class="st">"04_process_radtags/"</span>,</a>
<a class="sourceLine" id="cb15-5" data-line-number="5">  <span class="dt">parallel.core =</span> <span class="dv">12</span></a>
<a class="sourceLine" id="cb15-6" data-line-number="6">)</a></code></pre></div>
</div>
<div id="run_gstacks" class="section level2">
<h2 class="hasAnchor">
<a href="#run_gstacks" class="anchor"></a>run_gstacks</h2>
<p>For some reasons, if you used paired-end data, this step won’t work for now on macOS. It’s not a stackr problem, it’s a stacks problem. Please use linux … If you don’t have access to a linux computer, try using the cloud for this step <a href="https://thierrygosselin.github.io/RADseq_cloud_tutorial/radseq_cloud_tutorial.html">documentation</a>.</p>
<p>Function documentation: <code><a href="../reference/run_gstacks.html">?stackr::run_gstacks</a></code> (read it twice).</p>
<div id="single-end-1" class="section level3">
<h3 class="hasAnchor">
<a href="#single-end-1" class="anchor"></a>single-end</h3>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">g &lt;-<span class="st"> </span>stackr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/stackr/man/run_gstacks.html">run_gstacks</a></span>(</a>
<a class="sourceLine" id="cb16-2" data-line-number="2">  <span class="dt">M =</span> <span class="st">"02_project_info/population.map.turtle.catalog.tsv"</span>,</a>
<a class="sourceLine" id="cb16-3" data-line-number="3">  <span class="dt">parallel.core =</span> <span class="dv">12</span></a>
<a class="sourceLine" id="cb16-4" data-line-number="4">  )</a></code></pre></div>
</div>
<div id="paired-end-1" class="section level3">
<h3 class="hasAnchor">
<a href="#paired-end-1" class="anchor"></a>paired-end</h3>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">g &lt;-<span class="st"> </span>stackr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/stackr/man/run_gstacks.html">run_gstacks</a></span>(</a>
<a class="sourceLine" id="cb17-2" data-line-number="2">  <span class="dt">M =</span> <span class="st">"02_project_info/population.map.turtle.catalog.tsv"</span>,</a>
<a class="sourceLine" id="cb17-3" data-line-number="3">  <span class="dt">rm.pcr.duplicates =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb17-4" data-line-number="4">  <span class="dt">rm.unpaired.reads =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb17-5" data-line-number="5">  <span class="dt">unpaired =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb17-6" data-line-number="6">  <span class="dt">ignore.pe.reads =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb17-7" data-line-number="7">  <span class="dt">parallel.core =</span> <span class="dv">12</span></a>
<a class="sourceLine" id="cb17-8" data-line-number="8">)</a></code></pre></div>
</div>
</div>
<div id="run_populations" class="section level2">
<h2 class="hasAnchor">
<a href="#run_populations" class="anchor"></a>run_populations</h2>
<p>For this step, I like to keep it simple and get a VCF file. Filters are best handled in my package <a href="https://thierrygosselin.github.io/radiator/index.html">radiator</a>.</p>
<p>Function documentation: <code><a href="../reference/run_populations.html">?stackr::run_populations</a></code> (read it twice).</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">p &lt;-<span class="st"> </span>stackr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/stackr/man/run_populations.html">run_populations</a></span>(</a>
<a class="sourceLine" id="cb18-2" data-line-number="2">  <span class="dt">M =</span> <span class="st">"02_project_info/population.map.turtle.catalog.tsv"</span>, </a>
<a class="sourceLine" id="cb18-3" data-line-number="3">  <span class="dt">verbose =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb18-4" data-line-number="4">  <span class="dt">r =</span> <span class="fl">0.1</span>, </a>
<a class="sourceLine" id="cb18-5" data-line-number="5">  <span class="dt">R =</span> <span class="fl">0.1</span>,</a>
<a class="sourceLine" id="cb18-6" data-line-number="6">  <span class="dt">p =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb18-7" data-line-number="7">  <span class="dt">parallele.core =</span> <span class="dv">1</span>)</a></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#stackr-assumptions">stackr assumptions</a></li>
      <li><a href="#stackr-advantages">stackr advantages</a></li>
      <li><a href="#coding-style">Coding style</a></li>
      <li>
<a href="#workflow">Workflow</a><ul class="nav nav-pills nav-stacked">
<li><a href="#dependencies">Dependencies</a></li>
      <li><a href="#generate-project-folders">Generate project folders</a></li>
      <li><a href="#start-a-project">Start a project</a></li>
      <li><a href="#project-info-file">Project info file</a></li>
      <li><a href="#transfer-sequencing-files">Transfer sequencing files</a></li>
      <li><a href="#quality-control">Quality control</a></li>
      <li><a href="#run_process_radtags">run_process_radtags</a></li>
      <li><a href="#run_ustacks">run_ustacks</a></li>
      <li><a href="#run_cstacks">run_cstacks</a></li>
      <li><a href="#run_sstacks">run_sstacks</a></li>
      <li><a href="#run_tsv2bam">run_tsv2bam</a></li>
      <li><a href="#run_gstacks">run_gstacks</a></li>
      <li><a href="#run_populations">run_populations</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
