<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Run STACKS Version &gt;= 2.0 populations module — run_populations • stackr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Run STACKS Version &gt;= 2.0 populations module — run_populations" />
<meta property="og:description" content="Run STACKS
populations
module inside R!" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stackr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/stackr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/stackr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run STACKS Version &gt;= 2.0 populations module</h1>
    <small class="dont-index">Source: <a href='https://github.com/thierrygosselin/stackr/blob/master/R/run_populations.R'><code>R/run_populations.R</code></a></small>
    <div class="hidden name"><code>run_populations.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run <a href='http://catchenlab.life.illinois.edu/stacks/'>STACKS</a>
<a href='http://catchenlab.life.illinois.edu/stacks/comp/populations.php'>populations</a>
module inside R!</p>
    </div>

    <pre class="usage"><span class='fu'>run_populations</span>(
  P = <span class='st'>"06_ustacks_2_gstacks"</span>,
  V = <span class='kw'>NULL</span>,
  O = <span class='st'>"07_populations"</span>,
  <span class='kw'>M</span>,
  parallel.core = <span class='kw'>parallel</span>::<span class='fu'><a href='https://rdrr.io/r/parallel/detectCores.html'>detectCores</a></span>() <span class='op'>-</span> <span class='fl'>1</span>,
  batch.size = <span class='fl'>10000</span>,
  p = <span class='fl'>1</span>,
  r = <span class='fl'>0.3</span>,
  R = <span class='fl'>0.3</span>,
  H = <span class='fl'>TRUE</span>,
  min.maf = <span class='kw'>NULL</span>,
  min.mac = <span class='kw'>NULL</span>,
  max.obs.het = <span class='kw'>NULL</span>,
  write.single.snp = <span class='fl'>FALSE</span>,
  write.random.snp = <span class='fl'>FALSE</span>,
  B = <span class='kw'>NULL</span>,
  W = <span class='kw'>NULL</span>,
  e = <span class='kw'>NULL</span>,
  merge.sites = <span class='fl'>FALSE</span>,
  merge.prune.lim = <span class='kw'>NULL</span>,
  hwe = <span class='fl'>FALSE</span>,
  fstats = <span class='fl'>FALSE</span>,
  fst.correction = <span class='kw'>NULL</span>,
  p.value.cutoff = <span class='fl'>0.05</span>,
  k = <span class='fl'>FALSE</span>,
  smooth.fstats = <span class='fl'>FALSE</span>,
  smooth.popstats = <span class='fl'>FALSE</span>,
  sigma = <span class='fl'>150000</span>,
  bootstrap = <span class='fl'>FALSE</span>,
  N = <span class='fl'>100</span>,
  bootstrap.pifis = <span class='fl'>FALSE</span>,
  bootstrap.fst = <span class='fl'>FALSE</span>,
  bootstrap.div = <span class='fl'>FALSE</span>,
  bootstrap.phist = <span class='fl'>FALSE</span>,
  bootstrap.wl = <span class='kw'>NULL</span>,
  ordered.export = <span class='fl'>FALSE</span>,
  fasta.samples = <span class='fl'>TRUE</span>,
  fasta.samples_raw = <span class='fl'>FALSE</span>,
  fasta.loci = <span class='fl'>TRUE</span>,
  vcf = <span class='fl'>TRUE</span>,
  genepop = <span class='fl'>FALSE</span>,
  structure = <span class='fl'>FALSE</span>,
  radpainter = <span class='fl'>FALSE</span>,
  phase = <span class='fl'>FALSE</span>,
  fastphase = <span class='fl'>FALSE</span>,
  plink = <span class='fl'>FALSE</span>,
  hzar = <span class='fl'>FALSE</span>,
  phylip = <span class='fl'>FALSE</span>,
  phylip.var = <span class='fl'>FALSE</span>,
  phylip.var.all = <span class='fl'>FALSE</span>,
  treemix = <span class='fl'>FALSE</span>,
  no.hap.exports = <span class='fl'>FALSE</span>,
  h = <span class='fl'>FALSE</span>,
  verbose = <span class='fl'>FALSE</span>,
  v = <span class='fl'>FALSE</span>,
  log.fst.comp = <span class='fl'>FALSE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>P</th>
      <td><p>(path, character) Path to the directory containing all the STACKS files.
Default: <code>P = "06_ustacks_2_gstacks"</code>.</p></td>
    </tr>
    <tr>
      <th>V</th>
      <td><p>(character) Path to an input VCF file. When this module is used to
filter an existing vcf file.
Default: <code>V = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>O</th>
      <td><p>(character) Path to a directory where to write the output files.
With default: <code>O = "07_populations"</code>, the function creates a folder inside
<code>07_populations</code>, with date and time appended to <code>populations_</code>.</p></td>
    </tr>
    <tr>
      <th>M</th>
      <td><p>path to a population map file. The format is a tab-separated file,
with first column containing sample name and second column population id.
No heather (column name).
e.g. <code>M = "02_project_info/population.map.catalog.tsv"</code></p></td>
    </tr>
    <tr>
      <th>parallel.core</th>
      <td><p>(integer) enable parallel execution with num_threads threads.
Default: <code>parallel.core = parallel::detectCores() - 1</code></p></td>
    </tr>
    <tr>
      <th>batch.size</th>
      <td><p>(integer) The number of loci (de novo mode) or
chromosome (reference mode), to process in a batch.
Increase to speed analysis, uses more memory, decrease to save memory).
Default in de novo mode (loci/batch): <code>batch.size = 10000</code>.
Default in reference mode (chromosome/batch): <code>batch.size = 1</code>.</p></td>
    </tr>
    <tr>
      <th>p</th>
      <td><p>(integer) Minimum number of populations a locus must be present in to process a locus.
Default: <code>p = 1</code>.</p></td>
    </tr>
    <tr>
      <th>r</th>
      <td><p>(double) Minimum percentage of individuals in a population required to process a locus for that population.
Default: <code>r = 0.3</code>.</p></td>
    </tr>
    <tr>
      <th>R</th>
      <td><p>(double) Minimum percentage of individuals across populations required to process a locus.
Default: <code>r = 0.3</code>.</p></td>
    </tr>
    <tr>
      <th>H</th>
      <td><p>(logical) Apply the above filters haplotype wise
(unshared SNPs will be pruned to reduce haplotype-wise missing data).
Default: <code>H = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>min.maf</th>
      <td><p>(double) Specify a minimum minor allele frequency required to process a nucleotide site at a locus (0 &lt; min.maf &lt; 0.5).
Default: <code>min.maf = NULL</code>. Recommendation: use my other R package RADIATOR to filter data based on MAF.</p></td>
    </tr>
    <tr>
      <th>min.mac</th>
      <td><p>(integer) Specify a minimum minor allele count required to process a nucleotide site at a locus.
Default: <code>min.mac = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>max.obs.het</th>
      <td><p>(double) Specify a maximum observed heterozygosity required to process a nucleotide site at a locus.
Default: <code>max.obs.het = NULL</code>.  Recommendation: use my other R package RADIATOR to filter data based on heterozygosity.</p></td>
    </tr>
    <tr>
      <th>write.single.snp</th>
      <td><p>(logical) Restrict data analysis to only the first SNP per locus (implies --no-haps).
Default: <code>write.single.snp = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>write.random.snp</th>
      <td><p>(logical) Restrict data analysis to one random SNP per locus (implies --no-haps).
Default: <code>write.random.snp = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>B</th>
      <td><p>(character) Path to a file containing Blacklisted markers to be excluded from the export.
Default: <code>B = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>W</th>
      <td><p>(character) Path to a file containing Whitelisted markers to include in the export.
Default: <code>W = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>e</th>
      <td><p>(character) Restriction enzyme name.
Default: <code>e = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>merge.sites</th>
      <td><p>(logical) Merge loci that were produced from the same restriction enzyme cutsite (requires reference-aligned data).
Default: <code>merge.sites = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>merge.prune.lim</th>
      <td><p>(integer) When merging adjacent loci, if at least X<!-- % samples posses both loci prune the remaining samples out of the analysis. -->
Default: <code>merge.prune.lim = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>hwe</th>
      <td><p>(logical) Calculate divergence from Hardy-Weinberg equilibrium
using the exact test at the SNP level and Guo and Thompson MCMC algorithm at
the haplotype level.
Default: <code>hwe = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>fstats</th>
      <td><p>(logical) Enable SNP and haplotype-based F statistics.
Default: <code>fstats = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>fst.correction</th>
      <td><p>(character) Specify a correction to be applied to Fst values: 'p_value', 'bonferroni_win', or 'bonferroni_gen'.
Default: <code>fst.correction = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>p.value.cutoff</th>
      <td><p>(double) maximum p-value to keep an Fst measurement.
Also used as base for Bonferroni correction.
Default: <code>p.value.cutoff = 0.05</code>.</p></td>
    </tr>
    <tr>
      <th>k</th>
      <td><p>(logical) Enable kernel-smoothed Pi, Fis, Fst, Fst', and Phi_st calculations.
Default: <code>k = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>smooth.fstats</th>
      <td><p>(logical) Enable kernel-smoothed Fst, Fst', and Phi_st calculations.
Default: <code>smooth.fstats = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>smooth.popstats</th>
      <td><p>(logical) Enable kernel-smoothed Pi and Fis calculations.
Default: <code>smooth.popstats = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>sigma</th>
      <td><p>(integer) Standard deviation of the kernel smoothing weight distribution.
Default: <code>sigma = 150000</code> (150kb).</p></td>
    </tr>
    <tr>
      <th>bootstrap</th>
      <td><p>(logical) Turn on boostrap resampling for all smoothed statistics.
Default: <code>bootstrap = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>N</th>
      <td><p>(integer) Number of bootstrap resamplings to calculate.
Default: <code>N = 100</code>.</p></td>
    </tr>
    <tr>
      <th>bootstrap.pifis</th>
      <td><p>(logical) Turn on boostrap resampling for smoothed SNP-based Pi and Fis calculations.
Default: <code>bootstrap.pifis = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>bootstrap.fst</th>
      <td><p>(logical) Turn on boostrap resampling for smoothed Fst calculations based on pairwise population comparison of SNPs.
Default: <code>bootstrap.fst = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>bootstrap.div</th>
      <td><p>(logical) Turn on boostrap resampling for smoothed haplotype diveristy and gene diversity calculations based on haplotypes.
Default: <code>bootstrap.div = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>bootstrap.phist</th>
      <td><p>(logical) Turn on boostrap resampling for smoothed Phi_st calculations based on haplotypes.
Default: <code>bootstrap.phist = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>bootstrap.wl</th>
      <td><p>(character) Path to a whitelist file. Only use bootstrap loci contained in this whitelist.
Default: <code>bootstrap.wl = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>ordered.export</th>
      <td><p>(logical) If data is reference aligned, exports will be ordered; only a single representative of each overlapping site.
Default: <code>ordered.export = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>fasta.samples</th>
      <td><p>(logical) Output the sequences of the two haplotypes of each (diploid) sample, for each locus, in FASTA format.
stackr default is different than stacks default because I think having this info helps to identify problems.
Default: <code>fasta.samples = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>fasta.samples_raw</th>
      <td><p>(logical) Output all haplotypes observed in each sample, for each locus, in FASTA format.
Default: <code>fasta.samples_raw = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>fasta.loci</th>
      <td><p>(logical) Output consensus sequences of all loci, in FASTA format.
stackr default is different than stacks default because I think having this info helps to identify problems.
Default: <code>fasta.loci = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>vcf</th>
      <td><p>(logical) Output SNPs in Variant Call Format (VCF).
Default: <code>vcf = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>genepop</th>
      <td><p>(logical) Output results in GenePop format.
Default: <code>genepop = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>structure</th>
      <td><p>(logical) Output results in Structure format.
Default: <code>structure = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>radpainter</th>
      <td><p>(logical) Output results results in fineRADstructure/RADpainter format.
Default: <code>radpainter = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>phase</th>
      <td><p>(logical) Output genotypes in PHASE format.
Default: <code>phase = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>fastphase</th>
      <td><p>(logical) Output genotypes in fastPHASE format.
Default: <code>fastphase = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>plink</th>
      <td><p>(logical) Output genotypes in PLINK format.
Default: <code>plink = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>hzar</th>
      <td><p>(logical) Output genotypes in Hybrid Zone Analysis using R (HZAR) format.
Default: <code>hzar = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>phylip</th>
      <td><p>(logical) Output nucleotides that are fixed-within, and variant among populations in Phylip format for phylogenetic tree construction.
Default: <code>phylip = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>phylip.var</th>
      <td><p>(logical) Include variable sites in the phylip output encoded using IUPAC notation.
Default: <code>phylip.var = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>phylip.var.all</th>
      <td><p>(logical) Include all sequence as well as variable sites in the phylip output encoded using IUPAC notation.
Default: <code>phylip.var.all = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>treemix</th>
      <td><p>(logical) Output SNPs in a format useable for the TreeMix program (Pickrell and Pritchard).
Default: <code>treemix = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>no.hap.exports</th>
      <td><p>(logical) Omit haplotype outputs.
Default: <code>no.hap.exports = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>h</th>
      <td><p>Display this help messsage.
Default: <code>h = FALSE</code></p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>turn on additional logging.
Default: <code>verbose = FALSE</code></p></td>
    </tr>
    <tr>
      <th>v</th>
      <td><p>print program version.
Default: <code>v = FALSE</code></p></td>
    </tr>
    <tr>
      <th>log.fst.comp</th>
      <td><p>(logical) Log components of Fst/Phi_st calculations to a file.
Default: <code>log.fst.comp = FALSE</code></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><a href='http://catchenlab.life.illinois.edu/stacks/comp/populations.php'>populations</a>
returns different output depending on arguments selected.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Catchen JM, Amores A, Hohenlohe PA et al. (2011)
Stacks: Building and Genotyping Loci De Novo From Short-Read Sequences.
G3, 1, 171-182.</p>
<p>Catchen JM, Hohenlohe PA, Bassham S, Amores A, Cresko WA (2013)
Stacks: an analysis tool set for population genomics.
Molecular Ecology, 22, 3124-3140.</p>
<p>Guo SW, Thompson EA (1992)
Performing the exact test of Hardy-Weinberg proportion for multiple alleles.
Biometrics, 48, 361-372.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='http://catchenlab.life.illinois.edu/stacks/comp/populations.php'>populations</a>
<a href='http://catchenlab.life.illinois.edu/stacks/'>STACKS Version 2.0b</a></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>if</span> (<span class='fl'>FALSE</span>) {
<span class='kw'>pop</span> <span class='op'>&lt;-</span> <span class='kw'>stackr</span>::<span class='fu'>run_populations</span>(M = <span class='st'>"population.map.tsv"</span>)
}
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


