<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Run STACKS tsv2bam and merges BAM files — run_tsv2bam • stackr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Run STACKS tsv2bam and merges BAM files — run_tsv2bam" />
<meta property="og:description" content="Runs STACKS
tsv2bam
module and additionnally, this function will also generate a summary of
stacks tsv2bam and will merge in parallel BAM sample files into a unique
BAM catalog file using SAMtools or Sambamba.
tsv2bam
converts the data (single-end or paired-end) from being organized by sample
into being organized by locus. This allows downstream improvements
(e.g. Bayesian SNP calling)." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stackr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/stackr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/stackr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run STACKS tsv2bam and merges BAM files</h1>
    <small class="dont-index">Source: <a href='https://github.com/thierrygosselin/stackr/blob/master/R/run_tsv2bam.R'><code>R/run_tsv2bam.R</code></a></small>
    <div class="hidden name"><code>run_tsv2bam.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Runs <a href='http://catchenlab.life.illinois.edu/stacks/'>STACKS</a>
<a href='http://catchenlab.life.illinois.edu/stacks/stacks_v2.php'>tsv2bam</a>
module and additionnally, this function will also generate a summary of
stacks tsv2bam and will merge in parallel BAM sample files into a unique
BAM catalog file using SAMtools or Sambamba.
<a href='http://catchenlab.life.illinois.edu/stacks/stacks_v2.php'>tsv2bam</a>
converts the data (single-end or paired-end) from being organized by sample
into being organized by locus. This allows downstream improvements
(e.g. Bayesian SNP calling).</p>
    </div>

    <pre class="usage"><span class='fu'>run_tsv2bam</span>(
  <span class='kw'>P</span> <span class='kw'>=</span> <span class='st'>"06_ustacks_2_gstacks"</span>,
  <span class='kw'>M</span> <span class='kw'>=</span> <span class='st'>"02_project_info/population.map.tsv2bam.tsv"</span>,
  <span class='kw'>R</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>parallel.core</span> <span class='kw'>=</span> <span class='kw pkg'>parallel</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/r/parallel/detectCores.html'>detectCores</a></span>() - <span class='fl'>1</span>,
  <span class='kw'>cmd.path</span> <span class='kw'>=</span> <span class='st'>"/usr/local/bin/samtools"</span>,
  <span class='kw'>h</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>P</th>
      <td><p>(path, character) Path to the directory containing STACKS files.
Default: <code>P = "06_ustacks_2_gstacks"</code>.
Inside the folder, you should have:</p><ul>
<li><p><strong>the catalog files:</strong> starting with <code>batch_</code> and ending with
  <code>.alleles.tsv.gz, .snps.tsv.gz, .tags.tsv.gz</code>;</p></li>
<li><p><strong>3 files for each samples:</strong> The sample name is the prefix for
  the files ending with:
<code>.alleles.tsv.gz, .snps.tsv.gz, .tags.tsv.gz</code>.
Those files are created in the
<a href='http://catchenlab.life.illinois.edu/stacks/comp/ustacks.php'>ustacks</a>,
<a href='http://catchenlab.life.illinois.edu/stacks/comp/sstacks.php'>sstacks</a> and
<a href='http://catchenlab.life.illinois.edu/stacks/comp/cxstacks.php'>cxstacks</a>
modules.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>M</th>
      <td><p>(character, path) Path to a population map file.
Note that the <code>-s</code> option is not used inside <strong>stackr</strong>.
Default: <code>M = "02_project_info/population.map.tsv2bam.tsv"</code>.</p></td>
    </tr>
    <tr>
      <th>R</th>
      <td><p>(path, character) Directory where to find the paired-end reads files
(in fastq/fasta/bam (gz) format).</p></td>
    </tr>
    <tr>
      <th>parallel.core</th>
      <td><p>(integer) Enable parallel execution with the number of threads.
Default: <code>parallel.core = parallel::detectCores() - 1</code></p></td>
    </tr>
    <tr>
      <th>cmd.path</th>
      <td><p>(character, path) Provide the FULL path to SAMtools
program. See details on how to install SAMtools.
Default: <code>cmd.path = "/usr/local/bin/samtools"</code>.</p></td>
    </tr>
    <tr>
      <th>h</th>
      <td><p>Display this help messsage.
Default: <code>h = FALSE</code></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><a href='http://catchenlab.life.illinois.edu/stacks/stacks_v2.php'>tsv2bam</a>
returns a set of <code>.matches.bam</code> files.</p>
<p>The function <code>run_tsv2bam</code> returns a list with the number of individuals, the batch ID number,
a summary data frame and a plot containing:</p><ol>
<li><p>INDIVIDUALS: the sample id</p></li>
<li><p>ALL_LOCUS: the total number of locus for the individual (shown in subplot A)</p></li>
<li><p>LOCUS: the number of locus with a one-to-one relationship (shown in subplot B)
with the catalog</p></li>
<li><p>MATCH_PERCENT: the percentage of locus with a one-to-one relationship
with the catalog (shown in subplot C)</p>
<p>Addtionally, the function returns a catalog.bam file, generated
by merging all the individual BAM files in parallel.</p></li>
</ol>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><strong>Install SAMtools</strong>
<a href='link to detailed instructions on how to install SAMtools'>http://gbs-cloud-tutorial.readthedocs.io/en/latest/07_start_from_new_image.html#install-gbs-radseq-software-time-required-30min</a></p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Catchen JM, Amores A, Hohenlohe PA et al. (2011)
Stacks: Building and Genotyping Loci De Novo From Short-Read Sequences.
G3, 1, 171-182.</p>
<p>Catchen JM, Hohenlohe PA, Bassham S, Amores A, Cresko WA (2013)
Stacks: an analysis tool set for population genomics.
Molecular Ecology, 22, 3124-3140.</p>
<p>Li H., Handsaker B., Wysoker A., Fennell T., Ruan J., Homer N.,
Marth G., Abecasis G., Durbin R. and 1000 Genome Project Data Processing
Subgroup (2009) The Sequence alignment/map (SAM) format and SAMtools.
Bioinformatics, 25, 2078-9.</p>
<p>Li H A statistical framework for SNP calling,
mutation discovery, association mapping and population genetical parameter
estimation from sequencing data. Bioinformatics. 2011 Nov 1;27(21):2987-93.</p>
<p>A. Tarasov, A. J. Vilella, E. Cuppen, I. J. Nijman, and P. Prins.
Sambamba: fast processing of NGS alignment formats. Bioinformatics, 2015.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='http://catchenlab.life.illinois.edu/stacks/'>STACKS</a></p>
<p><a href='http://catchenlab.life.illinois.edu/stacks/stacks_v2.php'>stacks Version 2.0Beta6</a></p>
<p><a href='http://www.htslib.org'>SAMtools</a></p>
<p><a href='http://lomereiter.github.io/sambamba/'>Sambamba</a></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='co'># The simplest form of the function:</span>
<span class='no'>bam.sum</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>stackr</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/pkg/stackr/man/run_tsv2bam.html'>run_tsv2bam</a></span>() <span class='co'># that's it !</span>
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#references">References</a></li>
      <li><a href="#see-also">See also</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


