<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Run STACKS new module called gstacks — run_gstacks • stackr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Run STACKS new module called gstacks — run_gstacks" />
<meta property="og:description" content="Run STACKS
gstacks
module inside R!" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stackr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/stackr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/stackr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run STACKS new module called gstacks</h1>
    <small class="dont-index">Source: <a href='https://github.com/thierrygosselin/stackr/blob/master/R/run_gstacks.R'><code>R/run_gstacks.R</code></a></small>
    <div class="hidden name"><code>run_gstacks.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run <a href='http://catchenlab.life.illinois.edu/stacks/'>STACKS</a>
<a href='http://catchenlab.life.illinois.edu/stacks/stacks_v2.php'>gstacks</a>
module inside R!</p>
    </div>

    <pre class="usage"><span class='fu'>run_gstacks</span><span class='op'>(</span>
  denovo <span class='op'>=</span> <span class='cn'>TRUE</span>,
  P <span class='op'>=</span> <span class='st'>"06_ustacks_2_gstacks"</span>,
  M <span class='op'>=</span> <span class='st'>"02_project_info/population.map.tsv2bam.tsv"</span>,
  I <span class='op'>=</span> <span class='cn'>NULL</span>,
  B <span class='op'>=</span> <span class='cn'>NULL</span>,
  O <span class='op'>=</span> <span class='cn'>NULL</span>,
  parallel.core <span class='op'>=</span> <span class='fu'>parallel</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/r/parallel/detectCores.html'>detectCores</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>-</span> <span class='fl'>1</span>,
  model <span class='op'>=</span> <span class='st'>"marukilow"</span>,
  var.alpha <span class='op'>=</span> <span class='fl'>0.01</span>,
  gt.alpha <span class='op'>=</span> <span class='fl'>0.05</span>,
  rm.pcr.duplicates <span class='op'>=</span> <span class='cn'>FALSE</span>,
  rm.unpaired.reads <span class='op'>=</span> <span class='cn'>FALSE</span>,
  ignore.pe.reads <span class='op'>=</span> <span class='cn'>TRUE</span>,
  unpaired <span class='op'>=</span> <span class='cn'>TRUE</span>,
  kmer.length <span class='op'>=</span> <span class='fl'>31</span>,
  max.debruijn.reads <span class='op'>=</span> <span class='fl'>1000</span>,
  min.kmer.cov <span class='op'>=</span> <span class='fl'>2</span>,
  write.alignments <span class='op'>=</span> <span class='cn'>FALSE</span>,
  min.mapq <span class='op'>=</span> <span class='fl'>10</span>,
  max.clipped <span class='op'>=</span> <span class='fl'>0.2</span>,
  max.insert.len <span class='op'>=</span> <span class='fl'>1000</span>,
  details <span class='op'>=</span> <span class='cn'>FALSE</span>,
  phasing.cooccurrences.thr.range <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>2</span><span class='op'>)</span>,
  phasing.dont.prune.hets <span class='op'>=</span> <span class='cn'>FALSE</span>,
  h <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>denovo</th>
      <td><p>(logical) de novo vs reference-based mode.
Default: <code>denovo = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>P</th>
      <td><p>(path, character) De novo mode.
Path to the directory containing STACKS files.
Default: <code>P = "06_ustacks_2_gstacks"</code>.
Inside the folder, you should have:</p><ul>
<li><p><strong>the catalog files:</strong> starting with <code>batch_</code> and ending with
  <code>.alleles.tsv.gz, .snps.tsv.gz, .tags.tsv.gz and .bam</code>;</p></li>
<li><p><strong>5 files for each samples:</strong> The sample name is the prefix for
  the files ending with:
<code>.alleles.tsv.gz, .models.tsv.gz, .snps.tsv.gz, .tags.tsv.gz and .bam</code>.
Those files are created in the
<a href='http://catchenlab.life.illinois.edu/stacks/comp/ustacks.php'>ustacks</a>,
<a href='http://catchenlab.life.illinois.edu/stacks/comp/sstacks.php'>sstacks</a>,
<a href='http://catchenlab.life.illinois.edu/stacks/comp/cxstacks.php'>cxstacks</a>,
<a href='http://catchenlab.life.illinois.edu/stacks/stacks_v2.php'>tsv2bam</a>
modules</p></li>
</ul><p>.</p></td>
    </tr>
    <tr>
      <th>M</th>
      <td><p>(path, character) Path to a population map giving the list of samples.</p></td>
    </tr>
    <tr>
      <th>I</th>
      <td><p>(character, path) Reference-based mode.
Input directory containing BAM files.
Default: <code>I = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>B</th>
      <td><p>(character, path) Reference-based mode. Path to input BAM files.
The input BAM file(s) must be sorted by coordinate.
With <code>B</code>, records must be assigned to samples using BAM "reads groups"
(gstacks uses the ID/identifier and SM/sample name fields). Read groups
must be consistent if repeated different files. With <code>I</code>,
read groups are unneeded and ignored.
Please refer to the gstacks manual page for information. stacks also provide
information about how to
generate such a BAM file with Samtools or Sambamba, and examples.
Default: <code>B = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>O</th>
      <td><p>(character, path) Path to output directory. <code>stackr</code> by default
will output in the input directory.
Default: <code>O = NULL</code> (same as input).</p></td>
    </tr>
    <tr>
      <th>parallel.core</th>
      <td><p>(integer) Enable parallel execution with the number of threads.
Default: <code>parallel.core = parallel::detectCores() - 1</code>.</p></td>
    </tr>
    <tr>
      <th>model</th>
      <td><p>(character) The model to use to call variants and genotypes;
one of <code>"marukilow"</code>, <code>"marukihigh"</code>, or <code>"snp"</code>.
See ref for more details on algorithms.
Default: <code>model = "marukilow"</code>.</p></td>
    </tr>
    <tr>
      <th>var.alpha</th>
      <td><p>(double) Alpha threshold for discovering SNPs.
Default: <code>var.alpha = 0.01</code>.</p></td>
    </tr>
    <tr>
      <th>gt.alpha</th>
      <td><p>(double) Alpha threshold for calling genotypes.
Default: <code>gt.alpha = 0.05</code>.</p></td>
    </tr>
    <tr>
      <th>rm.pcr.duplicates</th>
      <td><p>(logical) remove all but one set of read pairs of
the same sample that have the same insert length
(implies <code>rm.unpaired.reads = TRUE</code>.).
Default: <code>rm.pcr.duplicates = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>rm.unpaired.reads</th>
      <td><p>(logical) Discard unpaired reads.
Default: <code>rm.unpaired.reads = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>ignore.pe.reads</th>
      <td><p>(logical) With default the function will
ignore paired-end reads even if present in the input.
Default: <code>ignore.pe.reads = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>unpaired</th>
      <td><p>(logical) Ignore read pairing (only for paired-end GBS;
treat READ2's as if they were READ1's).
Default: <code>unpaired = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>kmer.length</th>
      <td><p>(integer) De novo mode.
kmer length for the de Bruijn graph. For expert.
Default: <code>kmer.length = 31</code>.</p></td>
    </tr>
    <tr>
      <th>max.debruijn.reads</th>
      <td><p>(integer) Maximum number of reads to use in the
de Bruijn graph. For expert.
Default: <code>max.debruijn.reads = 1000</code>.</p></td>
    </tr>
    <tr>
      <th>min.kmer.cov</th>
      <td><p>(integer) De novo mode.
Minimum coverage to consider a kmer. For expert.
Default: <code>min.kmer.cov = 2</code>.</p></td>
    </tr>
    <tr>
      <th>write.alignments</th>
      <td><p>(logical) Save read alignments (heavy BAM files).
Default: <code>write.alignments = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>min.mapq</th>
      <td><p>(double) Reference-based mode.
Minimum PHRED-scaled mapping quality to consider a read.
Default: <code>min.mapq = 10</code>.</p></td>
    </tr>
    <tr>
      <th>max.clipped</th>
      <td><p>(double) Reference-based mode.
Maximum soft-clipping level, in fraction of read length.
Default: <code>max.clipped = 0.20</code>.</p></td>
    </tr>
    <tr>
      <th>max.insert.len</th>
      <td><p>(integer) Reference-based mode.
Maximum allowed sequencing insert length
Default: <code>max.insert.len = 1000</code>.</p></td>
    </tr>
    <tr>
      <th>details</th>
      <td><p>(logical) With default the function will write a more detailed
output.
Default: <code>details = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>phasing.cooccurrences.thr.range</th>
      <td><p>(integer) range of edge coverage thresholds to
iterate over when building the graph of allele cooccurrences for
SNP phasing.
Default: <code>phasing.cooccurrences.thr.range = c(1,2)</code>.</p></td>
    </tr>
    <tr>
      <th>phasing.dont.prune.hets</th>
      <td><p>(logical) Don't try to ignore dubious heterozygote
genotypes during phasing. By default, during phasing, dubious het are ignored.
Default: <code>phasing.dont.prune.hets = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>h</th>
      <td><p>Display this help messsage.
Default: <code>h = FALSE</code></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><a href='http://catchenlab.life.illinois.edu/stacks/stacks_v2.php'>tsv2bam</a>
returns a set of <code>.matches.bam</code> files.</p>
<p>The function <code>run_gstacks</code> returns a list with the number of individuals, the batch ID number,
a summary data frame and a plot containing:</p><ol>
<li><p>INDIVIDUALS: the sample id</p></li>
<li><p>ALL_LOCUS: the total number of locus for the individual (shown in subplot A)</p></li>
<li><p>LOCUS: the number of locus with a one-to-one relationship (shown in subplot B)
with the catalog</p></li>
<li><p>MATCH_PERCENT: the percentage of locus with a one-to-one relationship
with the catalog (shown in subplot C)</p>
<p>Addtionally, the function returns a batch_X.catalog.bam file that was generated
by merging all the individual BAM files in parallel.</p></li>
</ol>

    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Catchen JM, Amores A, Hohenlohe PA et al. (2011)
Stacks: Building and Genotyping Loci De Novo From Short-Read Sequences.
G3, 1, 171-182.</p>
<p>Catchen JM, Hohenlohe PA, Bassham S, Amores A, Cresko WA (2013)
Stacks: an analysis tool set for population genomics.
Molecular Ecology, 22, 3124-3140.</p>
<p>Maruki T, Lynch M (2017)
Genotype Calling from Population-Genomic Sequencing Data. G3, 7, 1393-1404.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='http://catchenlab.life.illinois.edu/stacks/'>STACKS</a></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='co'># The simplest form of the function with De novo data:</span>
<span class='va'>bam.sum</span> <span class='op'>&lt;-</span> <span class='fu'>stackr</span><span class='fu'>::</span><span class='fu'>run_gstacks</span><span class='op'>(</span><span class='op'>)</span> <span class='co'># that's it !</span>
<span class='co'># This will use, by default, the same population map used in run_tsv2bam</span>
<span class='op'>}</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


